File: Sources/Runtime/Microsoft.Psi.Interop/Rendezvous/RendezvousClient.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT license.

import socket, struct, threading
from enum import IntEnum

# Client which connects to a RendezvousServer and relays rendezvous information.
class RendezvousClient:
    PROTOCOL_VERSION = 2

    def __init__(self, host, port = 13331):
        self.serverAddress = (host, port)
        self.socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    def __sendByte(self, b):
        self.socket.send(struct.pack('b', b))

    def __readByte(self):
        b, = struct.unpack('b', self.socket.recv(1))
        return b

    def __sendInt(self, c):
        self.socket.send(struct.pack('<i', c))

    def __readInt(self):
        i, = struct.unpack('<i', self.socket.recv(4))
        return i

    def __sendString(self, s):
        e = s.encode()
        self.socket.send(struct.pack('b%ds' % len(e), len(e), e)) # assumes len < 128 (LEB128 encoded)
    
    def __readString(self):
        len = self.__readByte()
        buf = self.socket.recv(len)
        str, = struct.unpack('%ds' % len, buf)
        return str

    def __sendProtocolVersion(self):
        self.socket.send(struct.pack('<h', self.PROTOCOL_VERSION)) # protocol version

    def __readProtocolVersion(self):
        version, = struct.unpack('<h', self.socket.recv(2))
        if version != self.PROTOCOL_VERSION:
            raise Exception('RendezvousClient protocol mismatch %d' % self.PROTOCOL_VERSION);

    class Endpoint(IntEnum):
        TcpSource = 0
        NetMQSource = 1
        RemoteExporter = 2
        RemoteClockExporter = 3

    class TransportKind(IntEnum):
        Tcp = 0
        Udp = 1
        NamedPipes = 2

    def createStream(name, type):
        return { 'name': name,
                 'type': type }

    def createTcpEndpoint(host, port, streams):
        return { 'endpoint': RendezvousClient.Endpoint.TcpSource,
                 'host': host,
                 'port': port,
                 'streams': streams }

    def createNetMQEndpoint(address, streams):
        return { 'endpoint': RendezvousClient.Endpoint.NetMQSource,
                 'address': address,
                 'streams': streams }

    def createRemoteExporterEndpoint(host, port, transport, streams):
        return { 'endpoint': RendezvousClient.Endpoint.RemoteExporter,
                 'host': host,
                 'port': port,
                 'transport': transport,
                 'streams': streams }

    def createRemoteClockExporterEndpoint(host, port):
        return { 'endpoint': RendezvousClient.Endpoint.RemoteClockExporter,
                 'host': host,
                 'port': port,
                 'streams': [] }

    def createProcess(name, endpoints, version):
        return { 'name': name,
                 'endpoints': endpoints,
                 'version': version }

    def start(self, processAddedCallback = None, processRemovedCallback = None):
        self.socket.connect(self.serverAddress)
        self.onProcessAdded = processAddedCallback
        self.onProcessRemoved = processRemovedCallback
        self.__sendProtocolVersion()
        self.__readProtocolVersion()
        buf = self.socket.recv(1)
        len = struct.unpack('B')
        buf = self.socket.recv(len)
        addr, = struct.unpack('%ds' % len, buf)
        self.clientAddress = addr
        buf = self.socket.recv(4)
        numProcesses, = struct.unpack('<i', buf)
        for _ in range(numProcesses):
            self.__readProcessUpdate()
        self.thread = threading.Thread(target=self.__readProcessUpdates)
        self.thread.start()

    def stop(self):
        self.__sendByte(0) # disconnect
        self.socket.close()

    def addProcess(self, process):
        self.__sendByte(1) # add process
        self.__sendString(process['name'])
        self.__sendString(process['version'])
        self.__sendInt(len(process['endpoints']))
        for e in process['endpoints']:
            self.__sendByte(e['endpoint'])
            if e['endpoint'] == RendezvousClient.Endpoint.TcpSource:
                self.__sendString(e['host'])
                self.__sendInt(e['port'])
            elif e['endpoint'] == RendezvousClient.Endpoint.NetMQSource:
                self.__sendString(e['address'])
            elif e['endpoint'] == RendezvousClient.Endpoint.RemoteExporter:
                self.__sendString(e['host'])
                self.__sendInt(e['port'])
                self.__sendInt(e['transport'])
            elif e['endpoint'] == RendezvousClient.Endpoint.RemoteClockExporter:
                self.__sendString(e['host'])
                self.__sendInt(e['port'])
            else:
                raise Exception('Unknown type of Endpoint.')
            self.__sendInt(len(e['streams']))
            for s in e['streams']:
                self.__sendString(s['name'])
                self.__sendString(s['type'])

    def removeProcess(self, process):
        self.__sendByte(2) # remove process
        self.__sendString(process['name'])

    def __readStreams(self):
        streams = []
        count = self.__readInt()
        for _ in range(count):
            name = self.__readString()
            type = self.__readString()
            streams.append(RendezvousClient.createStream(name, type))
        return streams

    def __readEndpoint(self):
        endpoint = RendezvousClient.Endpoint(self.__readByte())
        if endpoint == RendezvousClient.Endpoint.TcpSource:
            host = self.__readString()
            port = self.__readInt()
            streams = self.__readStreams()
            return RendezvousClient.createTcpEndpoint(host, port, streams)
        elif endpoint == RendezvousClient.Endpoint.NetMQSource:
            address = self.__readString()
            streams = self.__readStreams()
            return RendezvousClient.createNetMQEndpoint(address, streams)
        elif endpoint == RendezvousClient.Endpoint.RemoteExporter:
            host = self.__readString()
            port = self.__readInt()
            transport = RendezvousClient.TransportKind(self.__readInt())
            streams = self.__readStreams()
            return RendezvousClient.createRemoteExporterEndpoint(host, port, transport, streams)
        elif endpoint == RendezvousClient.Endpoint.RemoteClockExporter:
            host = self.__readString()
            port = self.__readInt()
            self.__readInt() # stream count (zero)
            return RendezvousClient.createRemoteClockExporterEndpoint(host, port)
        else:
            raise Exception("Unknown type of Endpoint.")

    def __readProcess(self):
        name = self.__readString()
        version = self.__readString()
        numEndpoints = self.__readInt()
        endpoints = []
        for _ in range(numEndpoints):
            endpoints.append(self.__readEndpoint())
        return RendezvousClient.createProcess(name, endpoints, version)

    rendezvous = {}

    def __readProcessUpdate(self):
        update = self.__readByte()
        if update == 0: # disconnect
            return False
        elif update == 1: # add process
            process = self.__readProcess()
            self.rendezvous[process['name']] = process
            if self.onProcessAdded is not None:
                self.onProcessAdded(process)
            return True
        elif update == 2: # remove process
            name = self.__readString()
            removed = self.rendezvous.pop(name)
            if self.onProcessRemoved is not None:
                self.onProcessRemoved(removed)
            return True
        else:
            raise Exception("Unexpected rendezvous action.")

    def __readProcessUpdates(self):
        try:
            while self.__readProcessUpdate():
                pass
        except: # socket closed?
            pass

File: Sources/Runtime/Microsoft.Psi.Interop/Rendezvous/RendezvousClient.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT license.

namespace Microsoft.Psi.Interop.Rendezvous
{
    using System;
    using System.Diagnostics;
    using System.IO;
    using System.Net.Sockets;
    using System.Threading;

    /// <summary>
    /// Client which connects to a <see cref="RendezvousServer"/> and relays <see cref="Rendezvous"/> information.
    /// </summary>
    public class RendezvousClient : RendezvousRelay, IDisposable
    {
        private readonly string serverAddress;
        private readonly int port;
        private readonly EventWaitHandle connected = new (false, EventResetMode.ManualReset);

        private TcpClient client;
        private BinaryReader reader;
        private BinaryWriter writer;
        private bool active = false;
        private string clientAddress = null;

        /// <summary>
        /// Initializes a new instance of the <see cref="RendezvousClient"/> class.
        /// </summary>
        /// <param name="serverAddress">TCP address to which to connect.</param>
        /// <param name="port">Optional TCP port to which to connect.</param>
        /// <param name="rendezvous">Optional rendezvous instance to relay.</param>
        public RendezvousClient(string serverAddress, int port = RendezvousServer.DefaultPort, Rendezvous rendezvous = null)
            : base(rendezvous)
        {
            this.serverAddress = serverAddress;
            this.port = port;
            this.connected.Reset();
        }

        /// <summary>
        /// Gets wait handle for server connection being established.
        /// </summary>
        /// <remarks>This should be waited on prior to trusting the processes list.</remarks>
        public EventWaitHandle Connected => this.connected;

        /// <summary>
        /// Gets a value indicating whether the client is active.
        /// </summary>
        public bool IsActive => this.active;

        /// <summary>
        /// Gets the client address (available after connection established).
        /// </summary>
        public string ClientAddress => this.clientAddress;

        /// <summary>
        /// Start rendezvous client (blocking until connection is established).
        /// </summary>
        public void Start()
        {
            if (this.active)
            {
                throw new Exception($"{nameof(RendezvousClient)} already started.");
            }

            this.Rendezvous.ProcessAdded += this.ProcessAdded;
            this.Rendezvous.ProcessRemoved += this.ProcessRemoved;
            while (!this.active)
            {
                try
                {
                    (this.client = new TcpClient()).Connect(this.serverAddress, this.port);
                    var stream = this.client.GetStream();
                    this.reader = new BinaryReader(stream);
                    this.writer = new BinaryWriter(stream);
                    this.writer.Write(RendezvousServer.ProtocolVersion);
                    this.writer.Flush();
                    this.active = true;
                    new Thread(new ThreadStart(this.ReadFromServer)) { IsBackground = true }.Start();
                }
                catch (SocketException ex)
                {
                    Trace.WriteLine($"Failed to connect to {nameof(RendezvousServer)} (retrying): {ex.Message}");
                }
            }
        }

        /// <summary>
        /// Stop rendezvous client.
        /// </summary>
        public void Stop()
        {
            TryWriteDisconnect(this.writer);
            this.Rendezvous.ProcessAdded -= this.ProcessAdded;
            this.Rendezvous.ProcessRemoved -= this.ProcessRemoved;
            this.active = false;
            this.client?.Close();
            this.client?.Dispose();
            this.client = null;
            this.reader?.Dispose();
            this.reader = null;
            this.writer?.Dispose();
            this.writer = null;
        }

        /// <inheritdoc/>
        public void Dispose()
        {
            this.Stop();
            this.connected.Dispose();
        }

        private void ReadFromServer()
        {
            try
            {
                var version = this.reader.ReadInt16();
                if (version != RendezvousServer.ProtocolVersion)
                {
                    var ex = new IOException($"{nameof(RendezvousServer)} protocol mismatch ({version})");
                    this.ServerError(ex);
                    throw ex;
                }

                this.clientAddress = this.reader.ReadString();

                // initialize processes before signaling connected
                var count = this.reader.ReadInt32();
                for (var i = 0; i < count; i++)
                {
                    if (!this.ReadProcessUpdate(this.reader))
                    {
                        this.ServerError(new IOException($"{nameof(RendezvousServer)} disconnected."));
                    }
                }

                this.connected.Set();

                do
                {
                    if (!this.ReadProcessUpdate(this.reader))
                    {
                        this.ServerError(new IOException($"{nameof(RendezvousServer)} disconnected."));
                    }
                }
                while (this.active && this.client.Connected);
            }
            catch (Exception ex)
            {
                this.ServerError(ex);
                this.connected.Reset();
            }
        }

        private void NotifyServer(Rendezvous.Process process, Action<Rendezvous.Process, BinaryWriter> action)
        {
            try
            {
                action(process, this.writer);
            }
            catch (Exception ex)
            {
                this.ServerError(ex);
            }
        }

        private void ProcessAdded(object sender, Rendezvous.Process process)
        {
            if (this.writer != null)
            {
                this.NotifyServer(process, WriteAddProcess);
            }
        }

        private void ProcessRemoved(object sender, Rendezvous.Process process)
        {
            if (this.writer != null)
            {
                this.NotifyServer(process, WriteRemoveProcess);
            }
        }

        private void ServerError(Exception ex)
        {
            Trace.WriteLine($"{nameof(RendezvousServer)} error: {ex.Message}");
            this.Stop();
            this.OnError(ex);
        }
    }
}
